### PR TITLE
🚸 Add error handling for specific status codes

### PR DESCRIPTION
**Pull request by AuroraHuang22**
code 38 - nft is already bought
截圖 2023-06-07 下午11 39 10

code 13 - should show gas related error
截圖 2023-06-07 下午11 31 01

### CODE REVIEW

File: src/components/TxModal.vue [hunk 1]
```diff
@@ -308,6 +314,10 @@ export default {
           break;
         case 'insufficient':
         case 'failed':
+          logTrackerEvent(this, 'NFT', 'ClickRetry', this.classId, 1);
```
Comment (nwingt) [inline src/components/TxModal.vue:317]: I suggest to use more specific event name and use snake_case, e.g. `tx_modal_retry_button_clicked`

File: src/mixins/nft.js [hunk 2]
```diff
@@ -664,6 +664,17 @@ export default {
         this.classId,
         1
       );
+      const errorHandlingByCode = {
```
Comment (nwingt) [inline src/mixins/nft.js:667]: suggest to put this in a method 
```js
getErrorHandlerByCode(code) {
  switch (code) {
    case 38:
      return {
        message: 'NFT_IS_ALREADY_BOUGHT',
        runAction: () => this.updateNFTPurchaseInfo();
      };
       
    case 13: 
      return {
        message: 'INSUFFICIENT_GAS_FEE',
      };

    default:
      return {};
  }
}
```

File: src/mixins/nft.js [hunk 3]
```diff
@@ -731,7 +742,17 @@ export default {
...
-        if (code !== 0) throw new Error(`TX_FAILED_WITH_CODE_${code}`);
+        if (code !== 0) {
+          if (errorHandlingByCode[code]) {
+            this.uiSetTxError(errorHandlingByCode[code].message);
+            this.uiSetTxStatus(TX_STATUS.FAILED);
+            if (typeof errorHandlingByCode[code]?.action === 'function') {
+              errorHandlingByCode[code].action();
+            }
+            return undefined;
+          }
```
Comment (nwingt) [inline src/mixins/nft.js:753]:
```suggestion
          const errorHandler = getErrorHandlerByCode(code)
          if (errorHandler) {
            this.uiSetTxError(errorHandler.message);
            this.uiSetTxStatus(TX_STATUS.FAILED);
            if (typeof errorHandler?.runAction === 'function') {
              errorHandler.runAction();
            }
            return undefined;
          }
```

**Review by nwingt — APPROVED** [review]

File: src/locales/en.json [hunk 4]
```diff
@@ -579,6 +579,9 @@
   },
   "snackbar_error_buyLIKE": "Buy LIKE",
   "snackbar_error_insufficient": "Oops, Insufficient Balance.",
+  "snackbar_error_nft_is_already_bought": "This NFT has just been collected. Please try purchasing again",
```
Comment (williamchong) [inline src/locales/en.json:582]:
```suggestion
  "snackbar_error_nft_is_already_bought": "This NFT has just been collected by someone else. Please try purchasing again",

```

File: src/locales/zh-Hant.json [hunk 5]
```diff
@@ -569,6 +569,9 @@
   },
   "snackbar_error_buyLIKE": "購買 LIKE",
   "snackbar_error_insufficient": "噢，餘額不足！",
+  "snackbar_error_nft_is_already_bought": "此 NFT 剛被收藏了，請再重試購買",
```
Comment (williamchong) [inline src/locales/zh-Hant.json:572]:
```suggestion
  "snackbar_error_nft_is_already_bought": "此 NFT 剛被其他人收藏了，請再重試購買",

```

**Review by williamchong — CHANGES_REQUESTED** [review]

**Review by williamchong — APPROVED** [review]

**Pull request merged**